File: apps/ingestor/internal/event/geo.go
package event

import (
	"net/http"
	"strings"
)

// GeoInfo is derived from the client IP by a fronting CDN/proxy and passed as a
// header — so we read the header and never touch the raw IP for geo. region/city
// via a MaxMind GeoLite2 reader is a drop-in addition here.
type GeoInfo struct {
	CountryCode string // ISO-3166 alpha-2, or "" if unknown
	Region      string
	City        string
}

var countryHeaders = []string{
	"Cf-Ipcountry",             // Cloudflare
	"X-Vercel-Ip-Country",      // Vercel
	"X-Country-Code",           //
	"Fastly-Geoip-Countrycode", // Fastly
}

// ResolveGeo extracts geo data from request headers.
func ResolveGeo(h http.Header) GeoInfo {
	country := ""
	for _, key := range countryHeaders {
		v := h.Get(key)
		if len(v) == 2 && v != "XX" {
			country = strings.ToUpper(v)
			break
		}
	}
	return GeoInfo{
		CountryCode: country,
		Region:      h.Get("X-Vercel-Ip-Country-Region"),
		City:        h.Get("X-Vercel-Ip-City"),
	}
}

// ClientIP extracts the client IP from common proxy headers. Used transiently
// for the visitor hash only; never stored.
func ClientIP(h http.Header) string {
	if fwd := h.Get("X-Forwarded-For"); fwd != "" {
		first, _, _ := strings.Cut(fwd, ",")
		return strings.TrimSpace(first)
	}
	if ip := h.Get("Cf-Connecting-Ip"); ip != "" {
		return ip
	}
	if ip := h.Get("X-Real-Ip"); ip != "" {
		return ip
	}
	return "0.0.0.0"
}

File: apps/ingestor/src/lib/geo.ts
/**
 * Geo resolution. In production the ingestor runs behind a CDN/proxy (Cloudflare,
 * Vercel, Fastly) that already resolves the country from the client IP and passes
 * it as a header — so we read that and never touch the raw IP for geo.
 *
 * For finer region/city, plug in a MaxMind GeoLite2 reader here (gated on a
 * MAXMIND_DB_PATH env var); left out of the MVP to avoid the licensed .mmdb
 * download. The IP is used transiently for the visitor hash only, never stored.
 */

export interface GeoInfo {
	country_code: string; // ISO-3166 alpha-2, or '' if unknown
	region: string;
	city: string;
}

const COUNTRY_HEADERS = [
	'cf-ipcountry', // Cloudflare
	'x-vercel-ip-country', // Vercel
	'x-country-code',
	'fastly-geoip-countrycode'
];

export function resolveGeo(headers: Headers): GeoInfo {
	let country = '';
	for (const h of COUNTRY_HEADERS) {
		const v = headers.get(h);
		if (v && v.length === 2 && v !== 'XX') {
			country = v.toUpperCase();
			break;
		}
	}
	return {
		country_code: country,
		region: headers.get('x-vercel-ip-country-region') ?? '',
		city: headers.get('x-vercel-ip-city') ?? ''
	};
}
